src/c_api/c_api.cc

```diff
@@ -705,14 +705,15 @@ XGB_DLL int XGBoosterBoostOneIter(BoosterHandle handle,
                                   bst_float *grad,
                                   bst_float *hess,
                                   xgboost::bst_ulong len) {
-  std::vector<bst_gpair>& tmp_gpair = XGBAPIThreadLocalStore::Get()->tmp_gpair;
+  HostDeviceVector<bst_gpair>& tmp_gpair = XGBAPIThreadLocalStore::Get()->tmp_gpair;
   API_BEGIN();
   Booster* bst = static_cast<Booster*>(handle);
   std::shared_ptr<DMatrix>* dtr =
       static_cast<std::shared_ptr<DMatrix>*>(dtrain);
-  tmp_gpair.resize(len);
+  tmp_gpair.resize(len, -1);
```

Comment:
This resize function looks like it is assigning the value -1 if you are used to the std vector interface. Perhaps you can do it like:
```
void resize( size_type count, T value = T(), int device = -1);
```

Reply:
Done. Actually, I defined it as `resize(size_t n, int device = -1, T v = T())`, with device first, since I've got use cases where I want to have a non-default device, but I don't have any where I have a non-default fill value.

I've also added `T v = T()` parameter to the constructor.